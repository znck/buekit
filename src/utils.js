/**
 * Wrap to array.
 * @param {*} any
 * @returns {Array}
 */
export function toArray(any) {
	return Array.isArray(any) ? any : (
    any === undefined ? [] : [any]
  )
}

const isDot = /\./
export function normalizeProps (
  props,
  resolve = (name, suffix, value) => value,
  transform = (name, value) => value
) {
  const keys = Object.keys(props)
  const base = keys.filter(key => !isDot.test(key))
  const others = keys.filter(key => isDot.test(key))
  const normalized = {}

  base.forEach(
    name => {
      const start = name.length + 1
      normalized[name] = props[name] !== undefined ? transform(name, props[name]) : props[name]
      others.forEach(
        key => {
          if (props[key] !== undefined && key.startsWith(name)) {
            normalized[name] = [
              ...toArray(normalized[name]),
              ...toArray(resolve(name, key.substr(start), props[key]))
            ]
          }
        }
      )
    }
  )

  return normalized
}

/**
 * Merge vue data for functional components.
 * @param {object} local
 * @param {object} parent
 * @returns {object}
 */
export function mergeData(...nodes) {
	const classes = nodes.reduce(
    (result, node) => node && node.class ? result.concat(toArray(node.class)) : result, []
  )

	return Object.assign(...nodes, {class: classes})
}

/**
 * Set value and return target back.
 * @param {object} target
 * @param {string|number} key
 * @param {*} value
 * @returns {objext}
 */
export function fluentSet(target, key, value = true) {
	target[key] = value

	return target
}

/**
 * Convert list of keys to a hash map.
 * @param {Array<string|number>} keys
 * @returns {Object<string, bool>}
 */
export function toMap(keys) {
	return keys.reduce((result, key) => fluentSet(result, key), {})
}

export const responsiveTypes = toMap(['mobile', 'tablet', 'desktop', 'widescreen', 'fullhd'])

/**
 * Verify type is one of responsive types.
 * @param {string} typeName
 * @returns {bool}
 */
export function isResponsiveType(typeName) {
	if (typeName.indexOf('-') > -1) {
		typeName = typeName.split('-').pop()
	}

	return typeName in responsiveTypes
}

/**
 * Valid value with given validator.
 * @param {*} value
 * @param {function} validator
 * @returns {bool}
 */
export function validate(value, validator) {
	return toArray(value).every(validator)
}

export function isNumber(any) {
	return typeof any === 'number'
}

export function isString(any) {
	return typeof any === 'string'
}

export function isBoolean(any) {
	return typeof any === 'boolean'
}

export function isUndefined(any) {
	return typeof any === 'undefined'
}

export function isFunction(any) {
	return typeof any === 'function'
}

export function styleResolver(any) {
	any = any || new Proxy({}, { get: (_, key) => key })

	return key => any[key]
}

export const cssModule = styleResolver

export function isIgnoredVnode(vnode) {
	return !vnode || !vnode.tag || (vnode.isComment && !vnode.asyncFactory)
}

export function addStyleToVnode(vnode, classes) {
	if (isIgnoredVnode(vnode)) {
		return vnode
	}

	if (!vnode.data) {
		vnode.data = {}
	}

	vnode.data.class = vnode.data.class ? toArray(vnode.data.class).concat(toArray(classes)) : toArray(classes)

	return vnode
}

import flatten from 'lodash/flatten'

export const colors = ['primary', 'info', 'success', 'warning', 'danger', 'dark', 'light']

export function style(node, ...styles) {
  if (isIgnoredVnode(node)) return node

  node.data = {
    ...node.data,
    class: flatten([styles, toArray(node.data.class)]),
  }

  return node
}
