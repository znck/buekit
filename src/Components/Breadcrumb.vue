<script>
import PropTypes from '@znck/prop-types'
import { style, styleResolver, isIgnoredVnode } from '../utils'
import { createTag } from '../mixins'
import { sizes } from '../shared'


export default {
  name: 'Breadcrumb',
  functional: true,
  mixins: [createTag('nav')],
  props: {
    centered: PropTypes.bool.value(false),
    pushed: PropTypes.bool.value(false),
    separator: PropTypes.oneOf('succeeds', 'dot', 'bullet', 'arrow'),
    size: PropTypes.oneOf(sizes)
  },
  render (h, ctx) {
    const { size, tag, centered, pushed, separator } = ctx.props
    const _ = styleResolver(ctx.$style)
    const slots = ctx.slots()
    const links = (slots.default || []).filter(v => !isIgnoredVnode(v))
    const k = links.length - 1
    const wrapper = [
      h('ul', {}, links.map((link, i) => h('li', { class: [ i === k && _('is-active') ] }, [link])))
    ]

    return style(
      h(tag, ctx.data, wrapper),
      _('breadcrumb'),
      size && _(`is-${size}`),
      centered && _('is-centered'),
      pushed && _('is-right'),
      separator && _(`has-${separator}-separator`)
    )
  }
}
</script>

<style lang="scss" module>
@import '~bulma/sass/utilities/_all.sass';
@import '../_base.scss';
@import '~bulma/sass/components/breadcrumb.sass';
</style>
