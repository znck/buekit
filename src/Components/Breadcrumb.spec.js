import {shallow} from 'vue-test-utils'
import {createRenderer} from 'vue-server-renderer'

import Breadcrumb from './Breadcrumb.vue'

import { $style } from '../utils'

Breadcrumb.$style = $style

describe('Breadcrumb.vue', () => {
	it('matches snapshot', () => {
		const wrapper = shallow(Breadcrumb, { context: { props: {} } })
		const renderer = createRenderer()

		renderer.renderToString(wrapper.vm, (err, str) => {
			if (err) {
				throw new Error(err)
			}
			expect(str).toMatchSnapshot()
		})
	})
})
