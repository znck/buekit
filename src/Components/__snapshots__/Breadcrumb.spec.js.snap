// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Breadcrumb.vue matches snapshot 1`] = `
<nav class="breadcrumb breadcrumb">
    <ul></ul>
</nav>
`;
